### DISCUSSION TITLE
How to prevent "enter" from sending text to Copilot chat and instead create a new line?

### DISCUSSION THREAD

**Discussion by SamEdwardes** [community/community#86624] [Copilot Conversations] [answered]
### Select Topic Area

Product Feedback

### Body

When using Copilot chat, I often want to sent multiple lines of text. For example:

``````md
Can you please explain to me what this code will do:

```python
name = 'sam'
print(name)
```
``````

To create a new line, I need to use Shift+Enter. I would prefer if I could configure Enter to create a new line, and Shift+Enter to send the input.

**Comment by bonizario** [comment]
Here's a quick tutorial to achieve your desired behavior:

1. Press `CTRL+SHIFT+P` to open a VS Code panel that runs commands
2. Type: `Preferences: Open Keyboard Shortcuts`
3. Search for: `workbench.action.chat.acceptInput`
4. Right-click this command and click `Remove keybinding`

If this helped you, consider marking my answer as the "Accepted answer" 😄

**Reply by SamEdwardes** [reply]
Thank you so much! I will give this a try.

**Reply by SamEdwardes** [reply]
Thank you so much! I changed the binding to `SHIFT+ENTER`, it is much easier to type multiline inputs now :) 

![Screenshot 2024-01-18 at 08 01 13@2x](https://github.com/community/community/assets/18248949/01883f9c-bf8d-4fa5-96bf-bcb82299d645)

**Reply by SamEdwardes** [reply]
As of 2024-05-05 I think the name of the command has changed. It is now `workbench.action.chat.submit`. Change this keybinding to `shift + enter`.

Default:

![Screenshot 2024-05-05 at 16 27 26@2x](https://github.com/community/community/assets/18248949/5b74afaa-c256-47bb-94e3-d9dcfd7d0867)

Change:

![Screenshot 2024-05-05 at 16 28 58@2x](https://github.com/community/community/assets/18248949/19d3a533-e2d1-4f0e-b268-407553955e55)

**Reply by SamEdwardes** [reply, marked as answer]
Another update. I am now running this version of VS Code:

```
Version: 1.99.2
Commit: 4949701c880d4bdb949e3c0e6b400288da7f474b
Date: 2025-04-10T01:21:25.295Z
Electron: 34.3.2
ElectronBuildId: 11161073
Chromium: 132.0.6834.210
Node.js: 20.18.3
V8: 13.2.152.41-electron.0
OS: Darwin arm64 24.4.0
```

I have set up these bindings:

```jsonc
{
    // Disable the default `enter` submit behaviour for GitHub Copilot chat
    {
        "key": "enter",
        "command": "-workbench.action.chat.submit",
        "when": "chatInputHasText && chatRequestIsPaused && inChatInput && chatMode == 'ask' || chatInputHasText && inChatInput && !chatSessionRequestInProgress && chatMode == 'ask' || chatInstructionsAttached && chatRequestIsPaused && inChatInput && chatMode == 'ask' || chatInstructionsAttached && inChatInput && !chatSessionRequestInProgress && chatMode == 'ask'"
    },
    {
        "key": "cmd+enter",
        "command": "-workbench.action.chat.submitWithCodebase",
        "when": "chatInputHasText && chatRequestIsPaused && inChatInput || chatInputHasText && inChatInput && !chatSessionRequestInProgress || chatInstructionsAttached && chatRequestIsPaused && inChatInput || chatInstructionsAttached && inChatInput && !chatSessionRequestInProgress"
    },
    // Use the keyboard shortcuts you prefer.
    {
        "key": "cmd+enter",
        "command": "workbench.action.chat.submit",
        "when": "chatInputHasText && chatRequestIsPaused && inChatInput && chatMode == 'ask' || chatInputHasText && inChatInput && !chatSessionRequestInProgress && chatMode == 'ask' || chatInstructionsAttached && chatRequestIsPaused && inChatInput && chatMode == 'ask' || chatInstructionsAttached && inChatInput && !chatSessionRequestInProgress && chatMode == 'ask'"
    },
    {
        "key": "cmd+shift+enter",
        "command": "workbench.action.chat.submitWithCodebase",
        "when": "chatInputHasText && chatRequestIsPaused && inChatInput || chatInputHasText && inChatInput && !chatSessionRequestInProgress || chatInstructionsAttached && chatRequestIsPaused && inChatInput || chatInstructionsAttached && inChatInput && !chatSessionRequestInProgress"
    },
}
```

Feedback welcome if there is a more simple way to do this.

**Reply by InvictusNavarchus** [reply]
hii, this works for ask mode. what about for the new agent mode?

**Reply by NaeemAbdullahAkram** [reply]
> Another update. I am now running this version of VS Code:
> 
> ```
> Version: 1.99.2
> Commit: 4949701c880d4bdb949e3c0e6b400288da7f474b
> Date: 2025-04-10T01:21:25.295Z
> Electron: 34.3.2
> ElectronBuildId: 11161073
> Chromium: 132.0.6834.210
> Node.js: 20.18.3
> V8: 13.2.152.41-electron.0
> OS: Darwin arm64 24.4.0
> ```
> 
> I have set up these bindings:
> 
> ```json-comments
> {
>     // Disable the default `enter` submit behaviour for GitHub Copilot chat
>     {
>         "key": "enter",
>         "command": "-workbench.action.chat.submit",
>         "when": "chatInputHasText && chatRequestIsPaused && inChatInput && chatMode == 'ask' || chatInputHasText && inChatInput && !chatSessionRequestInProgress && chatMode == 'ask' || chatInstructionsAttached && chatRequestIsPaused && inChatInput && chatMode == 'ask' || chatInstructionsAttached && inChatInput && !chatSessionRequestInProgress && chatMode == 'ask'"
>     },
>     {
>         "key": "cmd+enter",
>         "command": "-workbench.action.chat.submitWithCodebase",
>         "when": "chatInputHasText && chatRequestIsPaused && inChatInput || chatInputHasText && inChatInput && !chatSessionRequestInProgress || chatInstructionsAttached && chatRequestIsPaused && inChatInput || chatInstructionsAttached && inChatInput && !chatSessionRequestInProgress"
>     },
>     // Use the keyboard shortcuts you prefer.
>     {
>         "key": "cmd+enter",
>         "command": "workbench.action.chat.submit",
>         "when": "chatInputHasText && chatRequestIsPaused && inChatInput && chatMode == 'ask' || chatInputHasText && inChatInput && !chatSessionRequestInProgress && chatMode == 'ask' || chatInstructionsAttached && chatRequestIsPaused && inChatInput && chatMode == 'ask' || chatInstructionsAttached && inChatInput && !chatSessionRequestInProgress && chatMode == 'ask'"
>     },
>     {
>         "key": "cmd+shift+enter",
>         "command": "workbench.action.chat.submitWithCodebase",
>         "when": "chatInputHasText && chatRequestIsPaused && inChatInput || chatInputHasText && inChatInput && !chatSessionRequestInProgress || chatInstructionsAttached && chatRequestIsPaused && inChatInput || chatInstructionsAttached && inChatInput && !chatSessionRequestInProgress"
>     },
> }
> ```
> 
> Feedback welcome if there is a more simple way to do this.

✅ Yes, successfully configured Copilot Chat to use Enter for new lines and Cmd+Enter to send.

keybindings disable the default submit behavior and reassign submission to Cmd+Enter and Cmd+Shift+Enter.

**Reply by NaeemAbdullahAkram** [reply]
✅ Yes, successfully configured Copilot Chat to use Enter for new lines and Cmd+Enter to send.

keybindings disable the default submit behavior and reassign submission to Cmd+Enter and Cmd+Shift+Enter.

**Reply by SamEdwardes** [reply]
> hii, this works for ask mode. what about for the new agent mode?

I found the same thing. There is an example that should work for all modes.

```json

[
    // Disable the default `enter` submit behaviour for GitHub Copilot chat
    {
        "key": "enter",
        "command": "-workbench.action.chat.submit",
        "when": "inChatInput && (chatInputHasText || chatInstructionsAttached)"
    },
    {
        "key": "cmd+enter",
        "command": "-workbench.action.chat.submitWithCodebase",
        "when": "inChatInput && (chatInputHasText || chatInstructionsAttached)"
    },
    // Setup new bindings for GitHub Copilot chat
    {
        "key": "cmd+enter",
        "command": "workbench.action.chat.submit",
        "when": "inChatInput && (chatInputHasText || chatInstructionsAttached)"
    },
    {
        "key": "cmd+shift+enter",
        "command": "workbench.action.chat.submitWithCodebase",
        "when": "inChatInput && (chatInputHasText || chatInstructionsAttached)"
    },
    {
        "key": "enter",
        "command": "editor.action.insertLineAfter",
        "when": "inChatInput"
    }
]
```

Courtesy of Jenny Brian (can't tag on mobile).

**Reply by JeremySeckinger** [reply]
> > hii, this works for ask mode. what about for the new agent mode?
> 
> I found the same thing. There is an example that should work for all modes.
> 
> ```json
> [
>     // Disable the default `enter` submit behaviour for GitHub Copilot chat
>     {
>         "key": "enter",
>         "command": "-workbench.action.chat.submit",
>         "when": "inChatInput && (chatInputHasText || chatInstructionsAttached)"
>     },
>     {
>         "key": "cmd+enter",
>         "command": "-workbench.action.chat.submitWithCodebase",
>         "when": "inChatInput && (chatInputHasText || chatInstructionsAttached)"
>     },
>     // Setup new bindings for GitHub Copilot chat
>     {
>         "key": "cmd+enter",
>         "command": "workbench.action.chat.submit",
>         "when": "inChatInput && (chatInputHasText || chatInstructionsAttached)"
>     },
>     {
>         "key": "cmd+shift+enter",
>         "command": "workbench.action.chat.submitWithCodebase",
>         "when": "inChatInput && (chatInputHasText || chatInstructionsAttached)"
>     },
>     {
>         "key": "enter",
>         "command": "editor.action.insertLineAfter",
>         "when": "inChatInput"
>     }
> ]
> ```
> 
> Courtesy of Jenny Brian (can't tag on mobile).

Thank you for this. Other answers weren't working.

**Reply by zacuo** [reply]
> > hii, this works for ask mode. what about for the new agent mode?
> 
> I found the same thing. There is ...

FYI. This fixes the issue where the suggest widget couldn't be selected, and also allows line breaks anywhere.
```jsonc
[
    // ============ Copilot Chat ============
    // Disable the default `enter` submit behaviour for GitHub Copilot chat
    {
        "key": "enter",
        "command": "-workbench.action.chat.submit",
        "when": "inChatInput && chatInputHasText && !chatSessionRequestInProgress && !withinEditSessionDiff"
    },
    {
        "key": "cmd+enter",
        "command": "-workbench.action.chat.submitWithCodebase",
        "when": "inChatInput && chatInputHasText && !chatSessionRequestInProgress"
    },
    // Setup new bindings for GitHub Copilot chat
    {
        "key": "cmd+enter",
        "command": "workbench.action.chat.submit",
        "when": "inChatInput && chatInputHasText && !chatSessionRequestInProgress && !withinEditSessionDiff"
    },
    {
        "key": "cmd+shift+enter",
        "command": "workbench.action.chat.submitWithCodebase",
        "when": "inChatInput && chatInputHasText && !chatSessionRequestInProgress"
    },
    {
        "key": "enter",
        "command": "",
        "when": "inChatInput && chatInputHasText && !suggestWidgetVisible"
    },
]
```

**Reply by GakEisenberg** [reply]
This should be the default. I'm convinced that these AI chats are configured this way so users 'accidentally' send requests and use up their usage allowances. Nearly ALL requests require newlines and we shouldn't have to constantly remember to use shift-enter for something that absolutely should not require it.

**Reply by ilazarev** [reply]
FYI: to insert line break at cursor position in gh chat : 
```json
{
    "key": "enter",
    "command": "type",
    "args": {
        "text": "\n"
    },
    "when": "inChatInput && !suggestWidgetVisible && (chatInputHasText || chatInstructionsAttached)"
 },
```

**Comment by lord-cyclone100** [comment]
Hello @SamEdwardes you can try out crtl+return/enter to go to the next line

**Comment by aspain-goodrx** [comment]
it seems the recent updates to copilot reset this for me, so I had to go back and redo it, but this still works as outlined above - thank you!

**Reply by SamEdwardes** [reply]
Same thing happened for me.

**Comment by NaeemAbdullahAkram** [comment]
✅ Yes,  successfully configured Copilot Chat to use Enter for new lines and Cmd+Enter to send.

keybindings disable the default submit behavior and reassign submission to Cmd+Enter and Cmd+Shift+Enter.

**Comment by davpaez** [comment]
Thank you @SamEdwardes , your solution worked! This lack of easy new line capability was driving me crazy.

Is there any way to vote or suggest that the default should be to NOT use `Enter` for submitting the prompt, instead leave it to create new lines?

I find weird that this is not the default.

To submit the prompt, the default should be `Ctrl+Enter`

**Comment by moezhx** [comment]
now it's `workbench.action.chat.submit`

**Comment by ApuEmdad** [comment]
it's not working in this version: 1.129.1

**Comment by lnhopkins** [comment]
Almost every month after updating VS Code they're changing it and I have to go to settings again and again set my key binding and this is driving me insane. Come on guys, make this setting persistent already!
